### PR TITLE
Forecasting model parallelization 

### PR DESCRIPTION
**Pull request by codeloop**
Forecasting model parallelization [ODSC-50382](https://jira.oci.oraclecorp.com/browse/ODSC-50382)
- make the model parallel using joblib
- allow object parallel access,
- some restructuring and formatting

### CODE REVIEW

**Review by ahosler — COMMENTED** [review]
image
Works well locally

**Review by ahosler — APPROVED** [review]
This looks awesome! Great work, looking forward to

File: ads/opctl/operator/lowcode/forecast/model/neuralprophet.py [hunk 1]
```diff
@@ -71,33 +72,31 @@ def __init__(self, config: ForecastOperatorConfig, datasets: ForecastDatasets):
...
+                boundaries = round((1 - self.spec.confidence_interval_width) / 2, 2)
+                self.quantiles = [boundaries, self.spec.confidence_interval_width + boundaries]
+
+            model_kwargs["quantiles"] = self.quantiles
+            self.forecast_output = ForecastOutput(
+                confidence_interval_width=self.spec.confidence_interval_width
+            )
 
-        for i, (target, df) in enumerate(full_data_dict.items()):
+            # for i, (target, df) in enumerate(full_data_dict.items()):
```
[the remark this answers is not in the document]
Comment (codeloop) [reply ads/opctl/operator/lowcode/forecast/model/neuralprophet.py:99]: Done

File: ads/opctl/operator/lowcode/forecast/operator_config.py [hunk 2]
```diff
@@ -132,6 +132,7 @@ def __post_init__(self):
             self.local_explanation_filename or "local_explanation.csv"
         )
         self.target_column = self.target_column or "Sales"
+        self.errors_dict_filename = "errors_dict.json"
```
[the remark this answers is not in the document]
Comment (codeloop) [reply ads/opctl/operator/lowcode/forecast/operator_config.py:135]: Done

File: ads/opctl/operator/lowcode/forecast/model/arima.py [hunk 3]
```diff
@@ -29,32 +30,32 @@ def __init__(self, config: ForecastOperatorConfig, datasets: ForecastDatasets):
...
-        if self.spec.confidence_interval_width is None:
-            self.spec.confidence_interval_width = 1 - self.spec.model_kwargs.get(
-                "alpha", 0.05
-            )
-        model_kwargs = self.spec.model_kwargs
-        model_kwargs["alpha"] = 1 - self.spec.confidence_interval_width
-        if "error_action" not in model_kwargs.keys():
-            model_kwargs["error_action"] = "ignore"
+    def _train_model(self, i, target, df):
+        # TODO : wrap it in try except and populate a errors_dict with the model name and the error
```
[the remark this answers is not in the document]
Comment (codeloop) [reply ads/opctl/operator/lowcode/forecast/model/arima.py]: Already done in the latest commit.

**Review by ahosler — APPROVED** [review]
This looks great! I tested with each framework and noticed very meaningful speedups.

**Pull request merged**